### PR TITLE
Fix various issues with the "make buildx" target

### PR DESCRIPTION
**Pull request by thaJeztah**
relates to https://github.com/moby/moby/pull/39340 (and slightly related to https://github.com/moby/moby/issues/40088)

See individual commit messages for a description 😅

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @cpuguy83 ptal

**Comment by thaJeztah** [comment]
Hm... interesting; am I dropping an env-var here, causing it to fall back to `go env` ?

```
 + make bundles/buildx
 mkdir bundles
 docker build -f ${BUILDX_DOCKERFILE:-Dockerfile.buildx} -t "moby-buildx:${BUILDX_COMMIT:-latest}" \
 	--build-arg BUILDX_COMMIT \
 	--build-arg BUILDX_REPO \
 	--build-arg GOOS=$(if [ -n "" ]; then echo ; else go env GOHOSTOS || uname | awk '{print tolower($0)}' || true; fi) \
 	--build-arg GOARCH=$(if [ -n "" ]; then echo ; else go env GOHOSTARCH || true; fi) \
 	. && \
 	id=$(docker create moby-buildx:${BUILDX_COMMIT:-latest}); \
 	if [ -n "${id}" ]; then docker cp ${id}:/usr/bin/buildx bundles/buildx && touch bundles/buildx; docker rm -f ${id}; fi \
 && bundles/buildx version
 /bin/sh: 4: go: not found
 /bin/sh: 5: go: not found
```

File: Makefile [hunk 1]
```diff
@@ -247,21 +247,19 @@ swagger-docs: ## preview the API documentation
...
+# because buildkit is not supported natively
 bundles/buildx: bundles ## build buildx CLI tool
-	# This intetionally is not using the `--output` flag from the docker CLI which is a buildkit option
-	# The idea here being that if buildx is being used, it's because buildkit is not supported natively
-	docker build -f $(BUILDX_DOCKERFILE) -t "moby-buildx:$(BUILDX_COMMIT)" \
+	docker build -f $${BUILDX_DOCKERFILE:-Dockerfile.buildx} -t "moby-buildx:$${BUILDX_COMMIT:-latest}" \
 		--build-arg BUILDX_COMMIT \
 		--build-arg BUILDX_REPO \
 		--build-arg GOOS=$$(if [ -n "$(GOOS)" ]; then echo $(GOOS); else go env GOHOSTOS || uname | awk '{print tolower($$0)}' || true; fi) \
 		--build-arg GOARCH=$$(if [ -n "$(GOARCH)" ]; then echo $(GOARCH); else go env GOHOSTARCH || true; fi) \
```
Comment (thaJeztah) [inline Makefile]: oh! `$(GOOS)` and `$(GOARCH)` are Make variables as well 🤔. I wonder why it worked before though

Comment (thaJeztah) [reply Makefile]: Perhaps I need to borrow this snippet from containerd; https://github.com/containerd/containerd/blob/036db34f37617d6a02f07b04fc2d35a91732c6e3/Makefile#L27-L52

Or set it in the Jenkinsfile

Comment (thaJeztah) [reply Makefile]: Ah! LOL.. then it worked, but didn't work because now we _actually_ used `v0.3.0`, which doesn't support s390x/ppc64le; The `copy` image use in v0.3.0 does not have s390x/ppc64le;

```
#3 docker-image://docker.io/tonistiigi/copy:v0.1.3@sha256:e57a3b4d6240f55ba...
#3 digest: sha256:150c626a5355601b2c71794d6e4a9af463e674e78c8f6cd2b0839485d62d8c5f
#3 name: "docker-image://docker.io/tonistiigi/copy:v0.1.3@sha256:e57a3b4d6240f55bac26b655d2cfb751f8b9412d6f7bb1f787e946391fb4b21b"
#3 started: 2019-10-14 16:57:58.920184883 +0000 UTC
#3 completed: 2019-10-14 16:57:58.920265067 +0000 UTC
#3 duration: 80.184
#3 started: 2019-10-14 16:57:58.920324378 +0000 UTC
#3 completed: 2019-10-14 16:57:58.922283274 +0000 UTC
#3 duration: 1.958896ms
#3 error: "no match for current platform linux/s390x in manifest sha256:e57a3b4d6240f55bac26b655d2cfb751f8b9412d6f7bb1f787e946391fb4b21b: not found"
#3 resolve docker.io/tonistiigi/copy:v0.1.3@sha256:e57a3b4d6240f55bac26b655d2cfb751f8b9412d6f7bb1f787e946391fb4b21b done
```

File: Makefile [hunk 2]
```diff
@@ -35,6 +35,49 @@ export VALIDATE_REPO
 export VALIDATE_BRANCH
 export VALIDATE_ORIGIN_BRANCH
 
+ifneq "$(strip $(shell command -v go 2>/dev/null))" ""
```
Comment (cpuguy83) [inline Makefile]: This all seems a bit overkill.
What are we trying to solve here?

Comment (thaJeztah) [reply Makefile]: @cpuguy83 I ran into this problem;

```
 /bin/sh: 4: go: not found
 /bin/sh: 5: go: not found
```

Comment (thaJeztah) [reply Makefile]: Likely because the `GOOS` and `GOARCH` make variables were not set

Comment (cpuguy83) [reply Makefile]: Yeah but doesn't what we already have solve that? And in the case no `go` the user just needs to set `GOOS` and `GOARCH` themselves.

Comment (thaJeztah) [reply Makefile]: But I just discovered what the actual problem was; we were now actually using buildx v0.3.0, so it tried to pull an image that doesn't exist for s390x/power. The `go not found` was (I think) just the output leaking from the `go env GOHOSTOS || uname` line; I'll remove that one commit

**Comment by thaJeztah** [comment]
Yo, Jenkins, where did you go?

File: Makefile (hunk 2 above)
Comment (thaJeztah) [reply Makefile]: Oh, actually.. (sleepy head); it's because s390x/power are now using buildkit to build the image 🤔

File: Dockerfile.buildx [hunk 3]
```diff
@@ -1,14 +1,27 @@
...
+ && export version="$(git describe --tags)" \
+ && export commit="$(git rev-parse --short HEAD)" \
+ && export ldflags="\
+    -X \"github.com/docker/buildx/version.Version=${version}\" \
+    -X \"github.com/docker/buildx/version.Revision=${commit}\" \
+    -X \"github.com/docker/buildx/version.Package=github.com/docker/buildx\"" \
+ && go build -ldflags "${ldflags}" -o /usr/bin/buildx ./cmd/buildx
+
+FROM golang:${GO_VERSION}-stretch
+COPY --from=build /usr/bin/buildx /usr/bin/buildx
```
Comment (thaJeztah) [inline Dockerfile.buildx]: Ah! 💡 I know why it failed in this branch; before this, there was no `COPY` in the Dockerfile, and because the `COPY` was added, it started downloading the `docker.io/tonistiigi/copy:v0.1.3` image. Apparently, that one is only available for amd64 ?

Comment (thaJeztah) [reply Dockerfile.buildx]: Yup; that one is only amd64, and armv7

File: Makefile (hunk 1 above)
Comment (thaJeztah) [reply Makefile]: See my other comment; it's because this PR adds a `COPY`, and therefore `docker.io/tonistiigi/copy:v0.1.3` has to be pulled, which is only available for `amd64` and `armv7`

**Comment by thaJeztah** [comment]
OK, this builds now, but I'll amend one commit message, and re-order commits (enabling BUILDX _before_ the commit that adds the `COPY`)

File: Dockerfile.buildx (hunk 3 above)
Comment (cpuguy83) [reply Dockerfile.buildx]: What's the reasoning for doing this second image?
This only produces a runnable image if the specified GOOS/GOARCH == the daemon's GOOS/GOARCH

In effect this is creating a 2nd image and taking even more space since buildx is duplicated.

Comment (thaJeztah) [reply Dockerfile.buildx]: The first stage won't be tagged, so can be garbage collected, and only the second one is preserved; the final image is roughly 870MB whereas without that stage, we're tagging an image that's 1.71GB; see the commit message https://github.com/moby/moby/commit/39f1a4ab44775428e95508ef8f186a81d60dc907

File: Dockerfile.buildx [hunk 4]
```diff
@@ -10,5 +10,12 @@ ARG GOOS
 ARG GOARCH
 # Keep these essentially no-op var settings for debug purposes.
 # It allows us to see what the GOOS/GOARCH that's being built for is.
-RUN GOOS=${GOOS} GOARCH=${GOARCH} go build -ldflags '-X github.com/docker/buildx/version.Version=${BUILDX_COMMIT} -X github.com/docker/buildx/version.Revision=${BUILDX_COMMIT} -X github.com/docker/buildx/version.Package=github.com/docker/buildx' -o /usr/bin/buildx ./cmd/buildx
-ENTRYPOINT ["/usr/bin/buildx"]
\ No newline at end of file
+RUN GOOS="${GOOS}" GOARCH="${GOARCH}" BUILDX_COMMIT="${BUILDX_COMMIT}" \
+ && export version="$(git describe --tags)" \
```
Comment (kolyshkin) [inline Dockerfile.buildx]: there should not be `&&` in here as what is before (var assignments) is not a command.

File: Dockerfile.buildx [hunk 5]
```diff
@@ -10,5 +10,12 @@ ARG GOOS
...
+ && export commit="$(git rev-parse --short HEAD)" \
```
Comment (kolyshkin) [inline Dockerfile.buildx]: 1 `export` allows for multiple var assignments
2 you don't need to export any variables as you're using them only here ("export" means make them available for child processes, and I don't see how it's necessary here)

Comment (kolyshkin) [reply Dockerfile.buildx]: Oh I see why you're using export... so you would use the vars later. Well, I still don't like the way it looks :(

Comment (thaJeztah) [reply Dockerfile.buildx]: Let me try again; I think I tried without export, and for some reason, things failed.

File: Makefile [hunk 6]
```diff
@@ -252,16 +252,19 @@ else
...
+export BUILDX_COMMIT
+
 bundles/buildx: BUILDX_DOCKERFILE ?= Dockerfile.buildx
-bundles/buildx: BUILDX_COMMIT ?= v0.3.0
 bundles/buildx: bundles ## build buildx CLI tool
 	# This intetionally is not using the `--output` flag from the docker CLI which is a buildkit option
 	# The idea here being that if buildx is being used, it's because buildkit is not supported natively
-	docker build -f $(BUILDX_DOCKERFILE) -t "moby-buildx:$(BUILDX_COMMIT)" \
+	docker build -f $(BUILDX_DOCKERFILE) -t "moby-buildx:$${BUILDX_COMMIT:-latest}" \
 		--build-arg BUILDX_COMMIT \
```
Comment (kolyshkin) [inline Makefile]: So, why not just use `--build-arg BUILDX_COMMIT=$(BUILDX_COMMIT)` instead?

File: Makefile [hunk 7]
```diff
@@ -252,16 +252,19 @@ else
...
 		--build-arg BUILDX_REPO \
 		--build-arg GOOS=$$(if [ -n "$(GOOS)" ]; then echo $(GOOS); else go env GOHOSTOS || uname | awk '{print tolower($$0)}' || true; fi) \
 		--build-arg GOARCH=$$(if [ -n "$(GOARCH)" ]; then echo $(GOARCH); else go env GOHOSTARCH || true; fi) \
 		. && \
-		id=$$(docker create moby-buildx:$(BUILDX_COMMIT)); \
-		if [ -n "$${id}" ]; then docker cp $${id}:/usr/bin/buildx $@ && touch $@; docker rm -f $${id}; fi
+		id=$$(docker create moby-buildx:$${BUILDX_COMMIT:-latest}); \
```
Comment (kolyshkin) [inline Makefile]: OK, so using `&&` from makefile is redundant. If any command fails, make stops right there. I guess this practice comes from Dockerfile `RUN` use, but makefile is different and if we can make things simpler, let's do it.

In other words,
```
	docker build ....
	id=$$(docker ...); if [ -n "$${id}" ]; ....
	$@ version
```

File: Makefile (hunk 6 above)
Comment (thaJeztah) [reply Makefile]: If possible, I like to have the build-args take their value from the environment. That way if no env-var is set, the value from the Dockerfile is leading. Putting the value _after_ means if it's not set, it will be `--build-arg=`, which is equivalent to `--build-arg=""`, which will override the value in the Dockerfile

File: Makefile (hunk 7 above)
Comment (thaJeztah) [reply Makefile]: Ok I rewrote it a bit; also put the `if` over a couple of lines, but let me know if that's too verbose.

File: Makefile [hunk 8]
```diff
@@ -247,21 +247,25 @@ swagger-docs: ## preview the API documentation
 
 .PHONY: buildx
 ifeq ($(BUILDX), bundles/buildx)
-buildx: bundles/buildx # build buildx cli tool
-else
-buildx:
+buildx: bundles/buildx ## build buildx cli tool
```
Comment (cpuguy83) [inline Makefile]: In https://github.com/moby/moby/pull/40091 buildx is a required pre-req for all builds.
What happens if we don't define `buildx:`?

Comment (thaJeztah) [reply Makefile]: It should be covered by the `.PHONY buildx`

```bash
mkdir foo && cd foo

echo ".PHONY: buildx" > Makefile

make buildx
# make: Nothing to be done for `buildx'.

echo $?
# 0

make nonexistingtarget
# make: *** No rule to make target `nonexistingtarget'.  Stop.

echo $?
2
```

File: Dockerfile.buildx [hunk 9]
```diff
@@ -10,5 +10,10 @@ ARG GOOS
...
 # It allows us to see what the GOOS/GOARCH that's being built for is.
-RUN GOOS=${GOOS} GOARCH=${GOARCH} go build -ldflags '-X github.com/docker/buildx/version.Version=${BUILDX_COMMIT} -X github.com/docker/buildx/version.Revision=${BUILDX_COMMIT} -X github.com/docker/buildx/version.Package=github.com/docker/buildx' -o /usr/bin/buildx ./cmd/buildx
-ENTRYPOINT ["/usr/bin/buildx"]
\ No newline at end of file
+RUN GOOS="${GOOS}" GOARCH="${GOARCH}" BUILDX_COMMIT="${BUILDX_COMMIT}" \
+    ldflags="\
+    -X \"github.com/docker/buildx/version.Version=$(git describe --tags)\" \
+    -X \"github.com/docker/buildx/version.Revision=$(git rev-parse --short HEAD)\" \
+    -X \"github.com/docker/buildx/version.Package=github.com/docker/buildx\"" \
+ && go build -ldflags "${ldflags}" -o /usr/bin/buildx ./cmd/buildx
```
Comment (kolyshkin) [inline Dockerfile.buildx]: `&&` looks artificial here as we don't have any command to succeed before it.

I also see a repetition of `github.com/docker/buildx` string; why don't we do this

```shell
  pkg="github.com/docker/buildx"; \
  ldflags="\
    -X \"${pkg}/version.Version=$(git describe --tags)\" \
    -X \"${pkg}/version.Revision=$(git rev-parse --short HEAD)\" \
    -X \"${pkg}/version.Package=${pkg}\""; \
  echo go build -ldflags "${ldflags}" -o /usr/bin/buildx ./cmd/buildx
```

I checked it works in both `bash` and `dash`, and even passes shellcheck

File: Makefile (hunk 6 above)
Comment (kolyshkin) [reply Makefile]: ok, makes sense

File: Dockerfile.buildx [hunk 10]
```diff
@@ -1,9 +1,12 @@
...
-FROM golang:${GO_VERSION}-stretch
+ARG BUILDX_COMMIT=v0.3.0
 ARG BUILDX_REPO=https://github.com/docker/buildx.git
+
+FROM golang:${GO_VERSION}-stretch
+ARG BUILDX_REPO
 RUN git clone "${BUILDX_REPO}" /buildx
 WORKDIR /buildx
-ARG BUILDX_COMMIT=master
+ARG BUILDX_COMMIT
```
Comment (kolyshkin) [inline Dockerfile.buildx]: Side note: docs at https://github.com/moby/moby/issues/37345#issuecomment-400245466 helped me to review this :)))

File: Dockerfile.buildx (hunk 9 above)
Comment (thaJeztah) [reply Dockerfile.buildx]: > && looks artificial here as we don't have any command to succeed before it.

arf. that's me messing up rebasing

> I also see a repetition of github.com/docker/buildx string; why don't we do this

👍 yup, I like that

File: Dockerfile.buildx (hunk 10 above)
Comment (thaJeztah) [reply Dockerfile.buildx]: Funny you reference it as "docs". I should try to find some time to incorporate it in the docs though

**Comment by thaJeztah** [comment]
Updated with the suggested changes

**Review by cpuguy83 — APPROVED** [review]
LGTM

File: Makefile [hunk 11]
```diff
@@ -252,16 +252,25 @@ else
...
 		--build-arg GOARCH=$$(if [ -n "$(GOARCH)" ]; then echo $(GOARCH); else go env GOHOSTARCH || true; fi) \
-		. && \
-		id=$$(docker create moby-buildx:$(BUILDX_COMMIT)); \
-		if [ -n "$${id}" ]; then docker cp $${id}:/usr/bin/buildx $@ && touch $@; docker rm -f $${id}; fi
+		.
+
+	id=$$(docker create moby-buildx:$${BUILDX_COMMIT:-latest}); \
+	if [ -n "$${id}" ]; then \
+		docker cp $${id}:/usr/bin/buildx $@; \
+		touch $@; \
```
Comment (kolyshkin) [inline Makefile]: OK, `&&` is still needed before `touch`, since otherwise we'll end up having an empty file in case `docker cp` fails, and this is not what we want here.

Comment (kolyshkin) [reply Makefile]: IOW,

`docker cp $${id}:/usr/bin/buildx $@ && touch $@; \`

otherwise all the same

Comment (kolyshkin) [reply Makefile]: in fact it's OK as it is since with empty file, `$@ version` below will fail anyway so up to you

File: Dockerfile.buildx [hunk 12]
```diff
@@ -10,5 +10,12 @@ ARG GOOS
...
 # It allows us to see what the GOOS/GOARCH that's being built for is.
-RUN GOOS=${GOOS} GOARCH=${GOARCH} go build -ldflags '-X github.com/docker/buildx/version.Version=${BUILDX_COMMIT} -X github.com/docker/buildx/version.Revision=${BUILDX_COMMIT} -X github.com/docker/buildx/version.Package=github.com/docker/buildx' -o /usr/bin/buildx ./cmd/buildx
-ENTRYPOINT ["/usr/bin/buildx"]
\ No newline at end of file
+RUN GOOS="${GOOS}" GOARCH="${GOARCH}" BUILDX_COMMIT="${BUILDX_COMMIT}"; \
+    pkg="github.com/docker/buildx"; \
+    ldflags="\
+      -X \"${pkg}/version.Version=$(git describe --tags)\" \
+      -X \"${pkg}/version.Revision=$(git rev-parse --short HEAD)\" \
+      -X \"${pkg}/version.Package=github.com/docker/buildx\" \
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply Dockerfile.buildx]: Ah, perhaps we should change to just `buildx`, so that it shows as;

```
buildx version
buildx v0.3.0 c967f1d
```

instead of

```
buildx version
github.com/docker/buildx v0.3.0 c967f1d
```

File: Makefile (hunk 11 above)
Comment (thaJeztah) [reply Makefile]: Ah, hm, yeah, overlooked that; I'll fix it to be correct

**Comment by thaJeztah** [comment]
@kolyshkin updated 👍

File: Dockerfile.buildx [hunk 13]
```diff
@@ -10,5 +10,12 @@ ARG GOOS
...
 # It allows us to see what the GOOS/GOARCH that's being built for is.
-RUN GOOS=${GOOS} GOARCH=${GOARCH} go build -ldflags '-X github.com/docker/buildx/version.Version=${BUILDX_COMMIT} -X github.com/docker/buildx/version.Revision=${BUILDX_COMMIT} -X github.com/docker/buildx/version.Package=github.com/docker/buildx' -o /usr/bin/buildx ./cmd/buildx
-ENTRYPOINT ["/usr/bin/buildx"]
\ No newline at end of file
+RUN GOOS="${GOOS}" GOARCH="${GOARCH}" BUILDX_COMMIT="${BUILDX_COMMIT}"; \
+    pkg="github.com/docker/buildx"; \
+    ldflags="\
+      -X \"${pkg}/version.Version=$(git describe --tags)\" \
+      -X \"${pkg}/version.Revision=$(git rev-parse --short HEAD)\" \
+      -X \"${pkg}/version.Package=buildx\" \
```
Comment (kolyshkin) [inline Dockerfile.buildx]: hmm. shouldn't it be `-X \"${pkg}/version.Package=${pkg}\" \`?

Comment (thaJeztah) [reply Dockerfile.buildx]: See https://github.com/moby/moby/pull/40089#discussion_r337160075

This way, it looks like

```
buildx version
buildx v0.3.0 c967f1d
```

instead of

```
buildx version
github.com/docker/buildx v0.3.0 c967f1d
```

**Comment by thaJeztah** [comment]
rebased to kick CI (master was broken because of a linting issue)

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
argh; failing on a flaky test;

```
 === Failed
[2019-10-21T23:02:53.486Z] === FAIL: amd64.integration-cli TestDockerSwarmSuite/TestAPISwarmNodeDrainPause (72.72s)
[2019-10-21T23:02:53.486Z]     --- FAIL: TestDockerSwarmSuite/TestAPISwarmNodeDrainPause (72.72s)
[2019-10-21T23:02:53.486Z]         daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerSwarmSuite/TestAPISwarmNodeDrainPause"
[2019-10-21T23:02:53.486Z]         daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerSwarmSuite/TestAPISwarmNodeDrainPause"
[2019-10-21T23:02:53.486Z]         check_test.go:358: [d48494135b48e] joining swarm manager [d1822cb98a9bc]@0.0.0.0:2477, swarm listen addr 0.0.0.0:2478
[2019-10-21T23:02:53.486Z]         docker_api_swarm_node_test.go:106: timeout hit after 1m0s: output: "b99585122cdb\ncdbb41970aad\n"
[2019-10-21T23:02:53.486Z]         daemon.go:83: assertion failed: error is not nil: exit status 1
```
